app/sessionStoreShutdown.js

```diff
@@ -189,6 +186,39 @@ app.on('before-quit', (e) => {
     clearInterval(sessionStateSaveInterval)
   }
   module.exports.initiateSessionStateSave()
+}
+
+app.on('before-quit', (e) => {
+  if (shuttingDown && sessionStateStoreComplete) {
+    return
+  }
+  e.preventDefault()
+
+  if (!appStore) {
+    appStore = require('../js/stores/appStore')
+  }
+
+  // prevent dialog that is displayed when you have active downloads
+  const downloads = downloadUtil.getActiveDownloads(appStore.getState())
+  let downloading = downloads.size > 0
+
+  if (downloading && !shuttingDown && !downloadWarningIsOpened) {
```

Comment:
this is problematic because attempting to quit when the dialog is already open will bypass this conditional and call `beforeQuit`

Reply:
I also don't think we should embed this logic inside sessionStoreShutdown. `sessionStoreShutdown` shouldn't care about downloads and this will become unwieldy if we start adding other checks. Actions won't work well because this requires a synchronous check, but we can provide a hook to register shutdown listeners that can be called inside the existing `before-quit` block and short-circuit before `shuttingDown = true` if any of the callbacks return false.

Reply:
perhaps `addShutdownListener`?